core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -951,15 +951,22 @@ class KafkaApis(val requestChannel: RequestChannel,
               new OffsetFetchResponse(Errors.NONE, authorizedPartitionData.asJava, header.apiVersion)
             }
           } else {
-            val (authorizedPartitions, unauthorizedPartitions) = offsetFetchRequest.partitions.asScala
-              .partition(authorizeTopicDescribe)
-            val (error, authorizedPartitionData) = coordinator.handleFetchOffsets(offsetFetchRequest.groupId,
-              Some(authorizedPartitions))
+            val (authorizedPartitions, unauthorizedPartitions) =
+              offsetFetchRequest.partitions.asScala.partition(authorizeTopicDescribe)
+            val (error, authorizedPartitionData) =
+              coordinator.handleFetchOffsets(offsetFetchRequest.groupId, Some(authorizedPartitions))
             if (error != Errors.NONE)
               offsetFetchRequest.getErrorResponse(error)
             else {
+              val authorizedAndValidatedPartitionData = authorizedPartitionData.map {
+                case (topicPartition, offsetInfo) =>
+                  if (metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition).isEmpty)
```

Comment:
I think the KIP does not explicitly specify this but during implementation we decided that if, in older version, there is a coordinator or group level error at partition level we should make it a top level error so, considering this from another angle, clients can rely on the top level error independent of the protocol version ([here](https://github.com/apache/kafka/blob/fc1cfe475e8ae8458d8ddf119ce18d0c64653a70/clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java#L113)). These errors don't make sense at or apply to specific partitions anyways.

To summarize, for the scenario of two partitions one of which gets an `UNKNOWN_PARTITION` error and the other a `GROUP_COORDINATOR_NOT_AVAILABLE` error, you are suggesting (please correct me if I misunderstood) to not return an error at the top level and have the same two partitions in the response with their corresponding errors.

The current implementation of the latest protocol version, however, returns a response with a `GROUP_COORDINATOR_NOT_AVAILABLE` error at the top level and no partitions.

To me, since we would probably want to retry fetching offsets in a scenario like this, from a client point of view the second option seems to be satisfactory. There might be scenarios for which they may not be the case (perhaps if the error is `GROUP_AUTHORIZATION_FAILED`?). Pinging @hachikuji again hoping he can weigh in and we can reach a verdict :)